{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define specific colors for each chromosome\n",
    "chromosome_colors = {\n",
    "    'chr1': '#2b2d42',  # Dark slate gray\n",
    "    'chr2': '#d35400',  # Darker orange\n",
    "    'chr3': '#4b0082',  # Dark indigo\n",
    "    'chr4': '#800000',  # Oxblood red\n",
    "    'chr5': '#9467bd',  # Muted purple\n",
    "    'chr6': '#1b4f72',  # Deep blue\n",
    "    'chr7': '#0072b2',  # Sanger blue\n",
    "    'chr8': '#e377c2',  # Soft pink\n",
    "    'chr9': '#bcbd22',  # Olive green\n",
    "    'chr10': '#7f7f7f', # Neutral gray\n",
    "    'chr11': '#2ca02c', # Green\n",
    "    'chr12': '#ff9f1c', # Rich amber\n",
    "    'chr13': '#d90429', # Bright red\n",
    "    'chr14': '#8d99ae', # Cool gray\n",
    "    'chr15': '#8c564b', # Brown\n",
    "    'chr16': '#264653', # Charcoal blue\n",
    "    'chr17': '#e9c46a', # Sandy yellow\n",
    "    'chr18': '#e76f51', # Coral\n",
    "    'chr19': '#6c757d', # Slate gray\n",
    "    'chr20': '#f4a261', # Light tan\n",
    "    'chr21': '#1d3557', # Dark blue\n",
    "    'chr22': '#ff6347', # Tomato red\n",
    "    'chrX': '#3a0ca3',  # Royal blue\n",
    "    'chrY': '#a8dadc'   # Light blue\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import gzip\n",
    "\n",
    "# Function to load gzipped TSV file\n",
    "def load_gzipped_tsv(file_path):\n",
    "    with gzip.open(file_path, 'rt') as f:\n",
    "        data = pd.read_csv(f, sep='\\t')\n",
    "    return data\n",
    "\n",
    "# Function to plot AMBER BAF\n",
    "def plot_amber_baf(patient_id, ax=None):\n",
    "    base_dir = '/lustre/scratch126/casm/team274sb/bw18/twins_sarcoma/purple/amber_out'\n",
    "    file_path = f\"{base_dir}/{patient_id}/{patient_id}.amber.baf.tsv.gz\"\n",
    "    data = load_gzipped_tsv(file_path)\n",
    "\n",
    "    # Filter out invalid data\n",
    "    filtered_df = data[data['tumorBAF'] != -1]\n",
    "\n",
    "    # Convert chromosome to a categorical type with a specific order\n",
    "    chromosomes = [f'chr{i}' for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "    filtered_df['chromosome'] = pd.Categorical(filtered_df['chromosome'], categories=chromosomes, ordered=True)\n",
    "\n",
    "    # Sort by chromosome and position\n",
    "    filtered_df = filtered_df.sort_values(['chromosome', 'position'])\n",
    "\n",
    "    # Create a new column for the combined x-axis positions\n",
    "    filtered_df['combined_position'] = np.nan\n",
    "\n",
    "    # Calculate the combined position for each chromosome\n",
    "    start = 0\n",
    "    chrom_midpoints = {}\n",
    "    for chrom in chromosomes:\n",
    "        chrom_data = filtered_df[filtered_df['chromosome'] == chrom]\n",
    "        end = start + len(chrom_data)\n",
    "        filtered_df.loc[filtered_df['chromosome'] == chrom, 'combined_position'] = range(start, end)\n",
    "        chrom_midpoints[chrom] = (start + end) // 2\n",
    "        start = end\n",
    "\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(20, 7))\n",
    "\n",
    "    # Plotting tumor BAF with alpha for points\n",
    "    for chrom in chromosomes:\n",
    "        chrom_data = filtered_df[filtered_df['chromosome'] == chrom]\n",
    "        ax.scatter(chrom_data['combined_position'], chrom_data['tumorBAF'], s=0.05, alpha=0.3, color=chromosome_colors[chrom])\n",
    "\n",
    "    # Plotting legend without alpha\n",
    "    handles = [plt.Line2D([0], [0], marker='o', color='w', markerfacecolor=chromosome_colors[chrom], markersize=10, label=chrom) for chrom in chromosomes]\n",
    "    ax.legend(handles=handles, loc='center left', bbox_to_anchor=(1, 0.5), fontsize='small', ncol=1, title=\"Chromosomes\", frameon=False)\n",
    "\n",
    "    ax.set_title(f'AMBER {patient_id} Tumor BAF', fontsize=24)  # Increased the font size for the title\n",
    "    ax.set_ylabel('B-Allele Frequency')\n",
    "    ax.set_xticks(list(chrom_midpoints.values()))\n",
    "    ax.set_xticklabels(list(chrom_midpoints.keys()), rotation='vertical')\n",
    "\n",
    "    if ax is None:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "def plot_cobalt_data(patient_id, ax=None):\n",
    "    base_dir = '/lustre/scratch126/casm/team274sb/bw18/twins_sarcoma/purple/cobalt_output'\n",
    "    file_path = f\"{base_dir}/{patient_id}/{patient_id}.cobalt.ratio.tsv\"\n",
    "    gz_file_path = f\"{file_path}.gz\"\n",
    "    \n",
    "    # Unzipping the file if necessary\n",
    "    if os.path.exists(gz_file_path):\n",
    "        with gzip.open(gz_file_path, 'rb') as f_in:\n",
    "            with open(file_path, 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "    \n",
    "    # Reading the TSV file\n",
    "    df = pd.read_csv(file_path, sep='\\t')\n",
    "    \n",
    "    # Filter out invalid data\n",
    "    filtered_df = df[df['tumorReadCount'] != -1]\n",
    "    \n",
    "    # Convert chromosome to a categorical type with a specific order\n",
    "    chromosomes = [f'chr{i}' for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "    filtered_df['chromosome'] = pd.Categorical(filtered_df['chromosome'], categories=chromosomes, ordered=True)\n",
    "    \n",
    "    # Sort by chromosome and position\n",
    "    filtered_df = filtered_df.sort_values(['chromosome', 'position'])\n",
    "    \n",
    "    # Create a new column for the combined x-axis positions\n",
    "    filtered_df['combined_position'] = np.nan\n",
    "    \n",
    "    # Calculate the combined position for each chromosome\n",
    "    start = 0\n",
    "    chrom_midpoints = {}\n",
    "    for chrom in chromosomes:\n",
    "        chrom_data = filtered_df[filtered_df['chromosome'] == chrom]\n",
    "        end = start + len(chrom_data)\n",
    "        filtered_df.loc[filtered_df['chromosome'] == chrom, 'combined_position'] = range(start, end)\n",
    "        chrom_midpoints[chrom] = (start + end) // 2\n",
    "        start = end\n",
    "    \n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(20, 7))\n",
    "\n",
    "    # Plotting tumor GC ratios\n",
    "    for chrom in chromosomes:\n",
    "        chrom_data = filtered_df[filtered_df['chromosome'] == chrom]\n",
    "        filtered_df.loc[filtered_df['chromosome'] == chrom, 'tumorGCRatio'] = np.clip(chrom_data['tumorGCRatio'], 0, 5)  # Clip values to a maximum of 5\n",
    "        ax.scatter(chrom_data['combined_position'], filtered_df.loc[filtered_df['chromosome'] == chrom, 'tumorGCRatio'], s=1, alpha=0.6, color=chromosome_colors[chrom], label=chrom if chrom_data.size > 0 else \"\")\n",
    "    ax.axhline(y=1, color='red', linestyle='--', linewidth=1)\n",
    "    ax.set_title(f'COBALT {patient_id} GC Normalised Tumor Reads', fontsize=24)  # Increased the font size for the title\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('GC Ratio')\n",
    "    ax.set_ylim(0, 5)  # Set y-axis limit for GC ratios\n",
    "    ax.set_xticks(list(chrom_midpoints.values()))\n",
    "    ax.set_xticklabels(list(chrom_midpoints.keys()), rotation='vertical')\n",
    "    \n",
    "    # Custom legend similar to AMBER\n",
    "    handles = [plt.Line2D([0], [0], marker='o', color='w', markerfacecolor=chromosome_colors[chrom], markersize=10, label=chrom) for chrom in chromosomes]\n",
    "    ax.legend(handles=handles, loc='center left', bbox_to_anchor=(1, 0.5), fontsize='small', ncol=1, title=\"Chromosomes\", frameon=False)\n",
    "\n",
    "    if ax is None:\n",
    "        plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def plot_allele_copy_numbers_segment(patient_id, ax=None):\n",
    "    base_dir = '/lustre/scratch126/casm/team274sb/bw18/twins_sarcoma/purple/purple_output'\n",
    "    file_path_segment = f\"{base_dir}/{patient_id}/without_GRIDDS/{patient_id}.purple.segment.tsv\"\n",
    "    file_path_purity = f\"{base_dir}/{patient_id}/without_GRIDDS/{patient_id}.purple.purity.tsv\"\n",
    "    \n",
    "    # Read the segment file\n",
    "    df_segment = pd.read_csv(file_path_segment, sep='\\t')\n",
    "    \n",
    "    # Ensure chromosomes are ordered\n",
    "    chromosomes = [f'chr{i}' for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "    df_segment['chromosome'] = pd.Categorical(df_segment['chromosome'], categories=chromosomes, ordered=True)\n",
    "    df_segment = df_segment.sort_values(['chromosome', 'start'])\n",
    "    \n",
    "    # Create a new column for the combined x-axis positions\n",
    "    df_segment['combined_position'] = np.nan\n",
    "    start = 0\n",
    "    chrom_midpoints = []\n",
    "    for chrom in chromosomes:\n",
    "        chrom_data = df_segment[df_segment['chromosome'] == chrom]\n",
    "        end = start + len(chrom_data)\n",
    "        df_segment.loc[df_segment['chromosome'] == chrom, 'combined_position'] = range(start, end)\n",
    "        chrom_midpoints.append((start + end) / 2)\n",
    "        start = end\n",
    "    \n",
    "    # Read the purity file\n",
    "    df_purity = pd.read_csv(file_path_purity, sep='\\t')\n",
    "    purity = df_purity['purity'].iloc[0] * 100\n",
    "    min_purity = df_purity['minPurity'].iloc[0] * 100\n",
    "    max_purity = df_purity['maxPurity'].iloc[0] * 100\n",
    "    ploidy = df_purity['ploidy'].iloc[0]\n",
    "\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(20, 7))\n",
    "\n",
    "    # Plot Minor Allele Copy Number\n",
    "    ax.scatter(df_segment['combined_position'], df_segment['minorAlleleCopyNumber'], color='#003366', s=1, label='Minor Allele Copy Number')\n",
    "    \n",
    "    # Plot Major Allele Copy Number\n",
    "    ax.scatter(df_segment['combined_position'], df_segment['tumorCopyNumber'], color='#800020', s=1, label='Total Copy Number')\n",
    "    \n",
    "    ax.set_xlabel('Chromosome')\n",
    "    ax.set_ylabel('Copy Number')\n",
    "    title = f'Copy Number Call of {patient_id}: Purity - {purity:.2f}% ({min_purity:.2f}-{max_purity:.2f}% range); Ploidy - {ploidy:.2f}'\n",
    "    ax.set_title(title, fontsize=24)  # Increased the font size for the title\n",
    "    \n",
    "    # Set y-axis limit\n",
    "    ax.set_ylim(0, 6)\n",
    "    \n",
    "    # Set x-ticks to chromosome names in the middle and oriented vertically\n",
    "    ax.set_xticks(chrom_midpoints)\n",
    "    ax.set_xticklabels(chromosomes, rotation='vertical')\n",
    "    \n",
    "    ax.legend()\n",
    "    if ax is None:\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def plot_allele_copy_numbers_smooth(patient_id, ax=None):\n",
    "    base_dir = '/lustre/scratch126/casm/team274sb/bw18/twins_sarcoma/purple/purple_output'\n",
    "    file_path_segment = f\"{base_dir}/{patient_id}/without_GRIDDS/{patient_id}.purple.cnv.somatic.tsv\"\n",
    "    file_path_purity = f\"{base_dir}/{patient_id}/without_GRIDDS/{patient_id}.purple.purity.tsv\"\n",
    "    \n",
    "    # Read the segment file\n",
    "    df_segment = pd.read_csv(file_path_segment, sep='\\t')\n",
    "    \n",
    "    # Ensure chromosomes are ordered\n",
    "    chromosomes = [f'chr{i}' for i in range(1, 23)] + ['chrX', 'chrY']\n",
    "    df_segment['chromosome'] = pd.Categorical(df_segment['chromosome'], categories=chromosomes, ordered=True)\n",
    "    df_segment = df_segment.sort_values(['chromosome', 'start'])\n",
    "    \n",
    "    # Round the copy number values\n",
    "    df_segment['minorAlleleCopyNumber'] = df_segment['minorAlleleCopyNumber'].round()\n",
    "    df_segment['copyNumber'] = df_segment['copyNumber'].round()  # Use copyNumber instead of majorAlleleCopyNumber\n",
    "    \n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(20, 7))\n",
    "    \n",
    "    # Plot Minor and Total Copy Number for each chromosome\n",
    "    cumulative_offset = 0\n",
    "    chromosome_boundaries = []\n",
    "    buffer_between_segments = 50000  # Small buffer between segments\n",
    "    buffer_between_chromosomes = 1000000  # Larger buffer between chromosomes\n",
    "    \n",
    "    for chrom in chromosomes:\n",
    "        chrom_data = df_segment[df_segment['chromosome'] == chrom]\n",
    "        if not chrom_data.empty:\n",
    "            chromosome_boundaries.append((chrom, cumulative_offset))\n",
    "            for index, row in chrom_data.iterrows():\n",
    "                start_pos = cumulative_offset + row['start']\n",
    "                end_pos = cumulative_offset + row['end']\n",
    "                ax.plot([start_pos, end_pos], [row['minorAlleleCopyNumber'], row['minorAlleleCopyNumber']], color='#003366', lw=4, label='Minor Allele Copy Number' if index == 0 and chrom == chromosomes[0] else \"\")\n",
    "                ax.plot([start_pos, end_pos], [row['copyNumber'], row['copyNumber']], color='#800020', lw=4, label='Total Copy Number' if index == 0 and chrom == chromosomes[0] else \"\")\n",
    "            cumulative_offset += chrom_data['end'].max() + buffer_between_chromosomes\n",
    "    \n",
    "    # Add vertical lines to mark chromosome boundaries\n",
    "    for chrom, boundary in chromosome_boundaries:\n",
    "        ax.axvline(x=boundary, color='grey', linestyle='--')\n",
    "    \n",
    "    # Read the purity file\n",
    "    df_purity = pd.read_csv(file_path_purity, sep='\\t')\n",
    "    purity = df_purity['purity'].iloc[0] * 100\n",
    "    min_purity = df_purity['minPurity'].iloc[0] * 100\n",
    "    max_purity = df_purity['maxPurity'].iloc[0] * 100\n",
    "    ploidy = df_purity['ploidy'].iloc[0]\n",
    "\n",
    "    ax.set_xlabel('Chromosome')\n",
    "    ax.set_ylabel('Copy Number')\n",
    "    title = f'Copy Number Call of {patient_id}: Purity - {purity:.2f}% ({min_purity:.2f}-{max_purity:.2f}% range); Ploidy - {ploidy:.2f}'\n",
    "    ax.set_title(title, fontsize=24)\n",
    "    \n",
    "    # Set y-axis limit\n",
    "    ax.set_ylim(0, 6)\n",
    "    \n",
    "    # Set x-ticks to chromosome names in the middle and oriented vertically\n",
    "    chrom_midpoints = [(boundary + (df_segment[df_segment['chromosome'] == chrom]['end'].max() / 2)) for chrom, boundary in chromosome_boundaries]\n",
    "    ax.set_xticks(chrom_midpoints)\n",
    "    ax.set_xticklabels([chrom for chrom, _ in chromosome_boundaries], rotation='vertical')\n",
    "    \n",
    "    # Add legend\n",
    "    ax.legend()\n",
    "    \n",
    "    if ax is None:\n",
    "        plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
